main/src/library/Fixpoint3/Ast/ExecutableRam.flix

```diff
@@ -0,0 +1,320 @@
+    }
+    ///
+    /// The EDB of a program.
+    ///
+    @Internal
+    pub type alias Facts = Map[RelSym, Vector[(Vector[Int64], Boxed)]]
+
+    ///
+    /// `ConstWrites` describes constants used for queries and the position they should be
+    /// written to. Explicitly for `(id1, id2, val)` we have that
+    /// `searchEnv[id1][id2] = val`.
+    ///
+    /// They come from rules with constants like `Head(x):- Body(4, x)`. We would here
+    /// a `ConstWrite` of `Vector#{(idOfBody, 0, 4)}`.
+    ///
+    @Internal
+    pub type alias ConstWrites = Vector[(Int32, Int32, Int64)]
+
+    ///
+    /// `WriteTuple` describes where constants for a context-specified RowVar should be
+    /// written to. Explicitly for `(indexInFrom, indexTo, indexInTo)` we have that for
+    /// the context-specified `RowVar`, `rv`, write
+    /// `searchEnv[indexTo][indexInTo] = rv[indexInFrom]`.
+    ///
+    /// They come from rules with constants like `Head(x):- Body1(x), Body2(y, x)`. We would here
+    /// a `WriteTuple` of `Vector#{(0, idOfBody2, 1)}`.
+    ///
+    @Internal
+    pub type alias WriteTuple = (Int32, Int32, Int32)
+
+    ///
+    /// `Arities` describes the arity-information for a program.
+    ///
+    /// For e.g. `Head(x):- Body(x, _, _)`. Assuming `Head` gets position `0` and `Body` gets position `1`
+    /// We would here have `Vector#{1, 3}`.
+    ///
+    type alias Arities = Vector[Int32]
+
+    /////////////////////////////////////////////////////////////////////////////
+    // RamStmt                                                                 //
+    /////////////////////////////////////////////////////////////////////////////
+
+    ///
+    /// `Insert(body)` executes the RelOp `body`.
+    ///
+    /// `MergeInto(posFrom, posTo, leq, isLatticenal)` merges the facts of the index at `posFrom`
+    /// into `posTo`. `isLatticenal` is describes whether lattice semantics should be used
+    /// and if so `leq` determines how to settle conflicts.
+    ///
+    /// `Swap(newPos, deltaPos)` swaps the facts of the indexes placed at `newPos` and
+    /// `deltaPos`.
+    ///
+    /// `Purge(pos)` removes all facts from the index at position `pos`.
+    ///
+    /// `Seq(body)` is sequence of statements to be executed sequentially.
+    ///
+    /// `Par(body)` is a sequence of statments to be executed in parallel.
+    ///
+    /// `Until(condition, body)` checks if `condition` is false and repeadetly executes
+    /// `body` until `condition` is `true`.
+    ///
+    /// `Comment(string)` is a comment for debugging.
+    ///
+    /// `EstimateJoinSize(indexPosition, savePosition, search)` estimates the size of the
+    /// index at `indexPosition` as described by `search` and save the result to
+    /// `savePosition`. For details see `JoinOptimizer.flix`.
+    ///
+    @Internal
+    pub enum RamStmt {
+        case Insert(RelOp)
+        case MergeInto(Int32, Int32, Boxed -> Boxed -> Boxed, Bool)
```

Comment:
you could consider making this bool a more telling enum type. should be close to free

Reply:
Jonathan means to replaced the boolean by a two-value enum